src/lib.rs

```diff
@@ -961,3 +963,100 @@ mod imp {
         }
     }
 }
+
+#[cfg(not(any(unix, windows)))]
+mod imp {
+    use std::io;
+    use std::sync::{Arc, Mutex};
+    use std::sync::atomic::{AtomicBool, Ordering};
+    use std::sync::mpsc::{self, SyncSender, Receiver};
+    use std::thread::{Builder, JoinHandle};
+    use std::process::Command;
+
+    #[derive(Debug)]
+    pub struct Client {
+        tx: SyncSender<()>,
+        rx: Mutex<Receiver<()>>,
+    }
+
+    #[derive(Debug)]
+    pub struct Acquired(());
+
+    impl Client {
+        pub fn new(limit: usize) -> io::Result<Client> {
+            let (tx, rx) = mpsc::sync_channel(limit);
+            for _ in 0..limit {
+                tx.send(()).unwrap();
+            }
+            Ok(Client {
+                tx,
+                rx: Mutex::new(rx),
+            })
+        }
+
+        pub unsafe fn open(_s: &str) -> Option<Client> {
+            None
+        }
+
+        pub fn acquire(&self) -> io::Result<Acquired> {
+            self.rx.lock().unwrap().recv().unwrap();
+            Ok(Acquired(()))
+        }
+
+        pub fn release(&self, _data: Option<&Acquired>) -> io::Result<()> {
+            self.tx.send(()).unwrap();
+            Ok(())
+        }
+
+        pub fn string_arg(&self) -> String {
+            panic!(
+                "On this platform there is no cross process jobserver support,
+                 so Client::configure is not supported."
+            );
+        }
+
+        pub fn configure(&self, _cmd: &mut Command) {
+            unreachable!();
+        }
+    }
+
+    #[derive(Debug)]
+    pub struct Helper {
+        thread: JoinHandle<()>,
+        quitting: Arc<AtomicBool>,
+    }
+
+    pub fn spawn_helper(client: ::Client,
+                        rx: Receiver<()>,
+                        mut f: Box<FnMut(io::Result<::Acquired>) + Send>)
+        -> io::Result<Helper>
+    {
+        let quitting = Arc::new(AtomicBool::new(false));
+        let quitting2 = quitting.clone();
+        let thread = Builder::new().spawn(move || {
+            'outer:
+            for () in rx {
+                loop {
+                    let res = client.acquire();
+                    if quitting2.load(Ordering::SeqCst) {
+                        break 'outer;
+                    }
+                    f(res);
+                    break;
+                }
+            }
+        })?;
+
+        Ok(Helper {
+            thread: thread,
+            quitting: quitting,
+        })
+    }
+
+    impl Helper {
+        pub fn join(self) {
+            self.quitting.store(true, Ordering::SeqCst);
```

Comment:
In addition to setting a quietting flag this will need to send a message on the channel to ensure that the other thread, if asleep, wakes up to see the flag.

Reply:
`HelperThread` seems to drop the `Receiver`, causing the for loop to stop, which means even setting `quitting` to true is not necessary.

Reply:
Ah good point, but due to clones of `jobserver::Client` I'm not sure that's a guarantee that can be relied on?

Reply:
Every time `Client::into_helper_thread` is called, a new helper thread is created. The `HelperThread` handle is not clonable, so dropping it will cause the only `Sender` to be dropped, causing the thread to stop.

Reply:
Can you test out this doesn't block forever then?

```rust
let x = Client::new(32);
let y = x.clone();
x.into_helper_thread().join();
```

Reply:
It works, do you want me to add a test?

```rust
#[test]
fn no_helper_deadlock() {
    let x = crate::Client::new(32).unwrap();
    let _y = x.clone();
    std::mem::drop(x.into_helper_thread(|_| {}).unwrap());
}
```

Reply:
Yeah that'd be great, thanks!

Reply:
Added it.